smart_card_connector_app/src/mock-chrome-api.js

```diff
@@ -0,0 +1,190 @@
+  /**
+   * @param {!goog.testing.MockControl} mockControl
+   * @param {!goog.testing.PropertyReplacer} propertyReplacer
+   */
+  constructor(mockControl, propertyReplacer) {
+    /** @type {Object.<string, Function>} */
+    this.listeners_ = {};
+    for (const event of EXTENSION_API_EVENTS) {
+      // Mock chrome.smartCardProviderPrivate.*.addListener()/removeListener().
+      // Cast is to work around the issue that the return type of
+      // createFunctionMock() is too generic (as it depends on the value of the
+      // optional second argument).
+      const mockedAddListener = /** @type {!goog.testing.StrictMock} */ (
+          mockControl.createFunctionMock(`${event}.addListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.addListener`,
+          mockedAddListener);
+      const mockedRemoveListener = /** @type {!goog.testing.StrictMock} */ (
+          mockControl.createFunctionMock(`${event}.removeListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.removeListener`,
+          mockedRemoveListener);
+
+      // Check that a listener will be added to the event and the same listener
+      // will be removed.
+      chrome.smartCardProviderPrivate[event].addListener(
+          goog.testing.mockmatchers.isFunction);
+      mockedAddListener.$once().$does((listener) => {
+        this.listeners_[event] = listener;
+      });
+
+      chrome.smartCardProviderPrivate[event].removeListener(
+          goog.testing.mockmatchers.isFunction);
+      mockedRemoveListener.$once().$does((listener) => {
+        assertEquals(
+            'Failed to remove listener', listener, this.listeners_[event]);
+        delete this.listeners_[event];
+      });
+    }
+
+    // Mock chrome.smartCardProviderPrivate.report*Result functions.
+    for (const reportFunction of EXTENSION_FUNCTIONS) {
+      // Cast is to work around the issue that the return type of
+      // createFunctionMock() is too generic (as it depends on the value of the
+      // optional second argument).
+      const mockedReport = /** @type {!goog.testing.StrictMock} */ (
+          mockControl.createFunctionMock(`${reportFunction}`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${reportFunction}`, mockedReport);
+    }
+
+    // Mock ResultCode enum.
+    for (const resultCode of RESULT_CODES) {
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.ResultCode.${resultCode}`,
+          resultCode);
+    }
+  }
+
+  /**
+   * Dispatches the specified chrome.smartCardProviderPrivate API event with
+   * `args`.
+   * @param {string} eventName
+   * @param  {...*} args
+   */
+  async dispatchEvent(eventName, ...args) {
+    assertEquals(
+        'No listener to dispatch event to', 'function',
+        typeof this.listeners_[eventName]);
+    await this.listeners_[eventName](...args);
```

Comment:
One thought about this `await`... Are we sure it's waiting for the correct thing? IIUC it waits for the function that's added via `addListener`, but is it guaranteed to be async function or to resolve after completing all its work?

I think your real intention here is to wait until the corresponding `report...()` method is called on the API, but that's probably better to be waited explicitly.

Reply:
you're welcome